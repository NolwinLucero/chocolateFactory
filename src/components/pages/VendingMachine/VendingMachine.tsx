import React, { useState } from 'react';
import styled from 'styled-components';
import { VendingMachineProducts } from '../../organisms/VendingMachineProducts/VendingMachineProducts';
import { VendingMachineBanner } from '../../atoms/VendingMachineBanner/VendingMachineBanner';
import { Payment } from '../../molecules/Payment/Payment';
import { ProductSchema } from '../../../schemas/product';
import { getLowestPrice, convertPrice } from '../../../logic/logic';

export const VendingMachine: React.FC = () => {
  const [credits, setCredits] = useState(0);
  const [products, setProducts] = useState<ProductSchema[]>([
    {
      id: 1,
      name: 'Caramel',
      price: '$3.10',
      stock: 3,
    },
    {
      id: 2,
      name: 'Hazelnut',
      price: '$2.50',
      stock: 3,
    },
    {
      id: 3,
      name: 'Organic Raw',
      price: '$2.00',
      stock: 3,
    },
  ]);

  const onCreditChangeHandler = (credit: string) => {
    const convertedCredit = convertPrice(credit);
    setCredits(credits + convertedCredit);

    if (credits + convertedCredit >= getLowestPrice(products)) {
      alert('You have enough credits to buy any available chocolates!');
    }
  };

  const OnBuyHandler = (e: React.FormEvent, id: number) => {
    const productToBuy = products.find(
      (product: ProductSchema) => product.id === id
    );

    if (productToBuy) {
      const price = parseFloat(productToBuy.price?.replace('$', ''));

      if (productToBuy.stock <= 0) {
        alert('Not enough stock');
        return;
      }

      if (credits < price) {
        alert('Not enough credits');
        return;
      } else {
        alert(`Enjoy your ${productToBuy.name} chocolate!`);
        setCredits(credits - price);

        const updatedProducts = products.map((p) =>
          p.id === productToBuy.id ? { ...p, stock: p.stock - 1 } : p
        );

        setProducts(updatedProducts);
      }
    }
  };

  return (
    <>
      <VendingMachineBanner />
      <StyledContainer>
        <StyledVendingMachineProductsContainer>
          <VendingMachineProducts
            products={products}
            credits={credits}
            OnBuyHandler={OnBuyHandler}
          />
        </StyledVendingMachineProductsContainer>
        <StyledPaymentContainer>
          <Payment
            credits={credits}
            onCreditChangeHandler={onCreditChangeHandler}
          />
        </StyledPaymentContainer>
      </StyledContainer>
    </>
  );
};

const StyledContainer = styled.div`
  display: flex;
  flex-direction: row;
  width: 100%;
  padding-top: 40px;
`;

const StyledPaymentContainer = styled.div`
  margin-right: 40px;
  width: 33.333%;
  background-color: purple;
`;

const StyledVendingMachineProductsContainer = styled.div`
  margin-left: 40px;
  width: 66.666%;
`;
