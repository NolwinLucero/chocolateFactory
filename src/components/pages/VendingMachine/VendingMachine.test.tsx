import React from 'react';
import ReactDOM from 'react-dom';
import { VendingMachine } from './VendingMachine';
import renderer from 'react-test-renderer';

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<VendingMachine></VendingMachine>, div);
  ReactDOM.unmountComponentAtNode(div);
});

it('matches snapshot', () => {
  const tree = renderer.create(<VendingMachine></VendingMachine>).toJSON();
  expect(tree).toMatchSnapshot();
});
