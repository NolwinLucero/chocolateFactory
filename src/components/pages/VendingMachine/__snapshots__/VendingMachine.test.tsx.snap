// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
Array [
  <h1>
     Brisbane's Chocolate Factory
  </h1>,
  <div>
    Allowed amounts: 10c, 20c, 50c, $1, $2
  </div>,
  <div
    className="sc-hKwDye fRBLEc"
  >
    <div
      className="sc-jRQBWg euDwfV"
    >
      <div
        className="sc-gsDKAQ buhTwb"
      >
        <div
          className="sc-bdvvtL bbdKiI"
        >
          <div>
            Caramel
          </div>
          <img
            alt="Caramel"
            src="https://www.tasteofhome.com/wp-content/uploads/2018/01/Gooey-Chocolate-Caramel-Bars_EXPS_CMZ18_10572_D10_26_5b-1.jpg?fit=200,200"
          />
          <div>
            Stock: 
            3
          </div>
          <div>
            $3.10
          </div>
          <button
            data-testid="1"
            disabled={true}
            id="1"
            onClick={[Function]}
          >
            Buy
          </button>
        </div>
      </div>
      <div
        className="sc-gsDKAQ buhTwb"
      >
        <div
          className="sc-bdvvtL bbdKiI"
        >
          <div>
            Hazelnut
          </div>
          <img
            alt="Hazelnut"
            src="https://www.tasteofhome.com/wp-content/uploads/2018/01/Gooey-Chocolate-Caramel-Bars_EXPS_CMZ18_10572_D10_26_5b-1.jpg?fit=200,200"
          />
          <div>
            Stock: 
            3
          </div>
          <div>
            $2.50
          </div>
          <button
            data-testid="2"
            disabled={true}
            id="2"
            onClick={[Function]}
          >
            Buy
          </button>
        </div>
      </div>
      <div
        className="sc-gsDKAQ buhTwb"
      >
        <div
          className="sc-bdvvtL bbdKiI"
        >
          <div>
            Organic Raw
          </div>
          <img
            alt="Organic Raw"
            src="https://www.tasteofhome.com/wp-content/uploads/2018/01/Gooey-Chocolate-Caramel-Bars_EXPS_CMZ18_10572_D10_26_5b-1.jpg?fit=200,200"
          />
          <div>
            Stock: 
            3
          </div>
          <div>
            $2.00
          </div>
          <button
            data-testid="3"
            disabled={true}
            id="3"
            onClick={[Function]}
          >
            Buy
          </button>
        </div>
      </div>
    </div>
    <div
      className="sc-eCImPb bMopAL"
    >
      <h1
        data-testid="credits"
      >
        Credits: $
        0.00
      </h1>
      <div
        className="sc-dkPtRN"
      >
        <form>
          <label>
            Input coin:
            <input
              data-testid="credit"
              onChange={[Function]}
              type="text"
            />
          </label>
          <button
            data-testid="addCredit"
            onClick={[Function]}
          >
            Drop
          </button>
        </form>
      </div>
    </div>
  </div>,
]
`;
